akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,314 @@
+ */
+final case class JoinSeedNodes(seedNodes: immutable.Seq[Address]) extends ClusterCommand {
+
+  /**
+   * Java API:
+   *
+   * Join the specified seed nodes without defining them in config.
+   * Especially useful from tests when Addresses are unknown before startup time.
+   *
+   * An actor system can only join a cluster once. Additional attempts will be ignored.
+   * When it has successfully joined it must be restarted to be able to join another
+   * cluster or to join the same cluster again.
+   *
+   * Creates a defensive copy of the list to ensure immutability.
+   */
+  def this(seedNodes: java.util.List[Address]) = this(Util.immutableSeq(seedNodes))
+
+}
+
+/**
+ * Send command to issue state transition to LEAVING for the node specified by 'address'.
+ * The member will go through the status changes [[MemberStatus]] `Leaving` (not published to
+ * subscribers) followed by [[MemberStatus]] `Exiting` and finally [[MemberStatus]] `Removed`.
+ *
+ * Note that this command can be issued to any member in the cluster, not necessarily the
+ * one that is leaving. The cluster extension, but not the actor system or JVM, of the
+ * leaving member will be shutdown after the leader has changed status of the member to
+ * Exiting. Thereafter the member will be removed from the cluster. Normally this is
+ * handled automatically, but in case of network failures during this process it might
+ * still be necessary to set the node’s status to Down in order to complete the removal.
+ */
+final case class Leave(address: Address) extends ClusterCommand
+
+/**
+ * Send command to DOWN the node specified by 'address'.
+ *
+ * When a member is considered by the failure detector to be unreachable the leader is not
+ * allowed to perform its duties, such as changing status of new joining members to 'Up'.
+ * The status of the unreachable member must be changed to 'Down', which can be done with
+ * this method.
+ */
+final case class Down(address: Address) extends ClusterCommand
+
+/**
+ * Akka Typed Cluster API entry point
+ */
+object Cluster extends ExtensionId[Cluster] {
+
+  def createExtension(system: ActorSystem[_]): Cluster = new AdapterClusterImpl(system)
+
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] object AdapterClusterImpl {
+
+  private def subscriptionsBehavior(adaptedCluster: akka.cluster.Cluster) = Actor.deferred[ClusterStateSubscription] { ctx ⇒
+    var selfSeenUp = false
+    var upSubscribers: List[ActorRef[SelfUp]] = Nil
+    var removedSubscribers: List[ActorRef[SelfRemoved.type]] = Nil
+
+    adaptedCluster.subscribe(ctx.self.toUntyped, ClusterEvent.initialStateAsEvents, classOf[MemberEvent])
+
+    // important to not eagerly refer to it or we get a cycle here
+    def cluster = Cluster(ctx.system)
+    def updateCachedSelfMember(member: Member): Unit = {
+      // safe because we are the only writer
+      cluster.asInstanceOf[AdapterClusterImpl]._cachedSelfMember = member
```

Comment:
That means that `AdapterClusterImpl. _cachedSelfMember` and `AdapterClusterImpl.state` are not always in sync. Wonder if that can be confusing?

Perhaps we could introduce this `_cachedSelfMember` into untyped Cluster itself, and make sure it's updated together with _state. One way to implement it is to set `_cachedSelfMember=null` before `_state` is updated, and have a `def selfMember` that does the search in `_state` if `_cachedSelfMember == null`. Kind of our own lazy val.

Reply:
I realised we have a self in `readView` already, so optimising that according to these lines and then just accessing it from here. Trying to update it from `eventBusListener` on all self-related member events, but can't figure out how I get it to ever change to `Removed`.

Reply:
not sure about Removed, but perhaps hook in to the `shutdown` method?

Reply:
That was the hint I needed :)